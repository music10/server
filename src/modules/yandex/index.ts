export * from './yandex.module';
export * from './yandex.service';
