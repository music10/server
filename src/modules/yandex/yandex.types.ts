export enum Type {
  'playlist' = 'playlist',
  'artist' = 'artist',
}
