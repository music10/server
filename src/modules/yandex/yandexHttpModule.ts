import { HttpModule } from '@nestjs/axios';

/**
 * Http module for Yandex.Music API
 */
export const YandexHttpModule = HttpModule.registerAsync({
  useFactory: () => ({
    timeout: 5000,
    maxRedirects: 5,
    baseURL: 'https://api.music.yandex.net/',
    headers: {
      'Accept-Language': 'ru',
      'User-Agent': 'Musiq',
    },
  }),
});
