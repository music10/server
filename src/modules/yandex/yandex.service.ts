import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';
import { parseStringPromise } from 'xml2js';

import { PlaylistDto, TrackDto } from '../../dtos';
import { randomIntByMax } from '../../utils';
import { Type } from './yandex.types';

/**
 * Page size for search
 */
const PAGE_SIZE = 20;

/**
 * Owner ID for random playlists
 */
const USER_ID_FOR_RANDOM = 'music-blog';

/**
 * Service for Yandex.Music Api
 */
@Injectable()
export class YandexService {
  /**
   * SpotifyApiService constructor
   */
  constructor(private httpService: HttpService) {}

  /**
   * Get playlists
   */
  async getCherryPickPlaylists(): Promise<PlaylistDto[]> {
    return firstValueFrom(
      this.httpService.get('https://musiq.dergunov.net/cherry-pick.json'),
    ).then(({ data }) => data);
  }

  /**
   * Search playlists by query-string
   * @param {string} text
   * @param {Type} type
   * @return {PlaylistDto[]} playlists
   */
  async searchPlaylists(text: string, type: Type): Promise<PlaylistDto[]> {
    return firstValueFrom(
      this.httpService.get(`/search`, {
        params: {
          text,
          type,
          page: 0,
          page_size: PAGE_SIZE,
          nocorrect: false,
        },
      }),
    ).then(({ data }) => {
      if (type === Type.playlist) {
        return data.result?.playlists?.results
          .filter((p) => p.revision && p.trackCount > 40)
          .map((playlist) => ({
            id: playlist.playlistUuid,
            name: playlist.title,
            type: Type.playlist,
            cover: 'https://' + playlist?.cover?.uri?.replace('%%', '200x200'),
            url: `https://music.yandex.ru/users/${playlist?.owner?.login}/playlists/${playlist.kind}`,
          }));
      }
      if (type === Type.artist) {
        return data.result?.artists?.results
          .filter((p) => p.counts.tracks > 40)
          .map((artist) => ({
            id: artist.id,
            name: artist.name,
            type: Type.artist,
            cover: 'https://' + artist?.cover?.uri?.replace('%%', '200x200'),
            url: `https://music.yandex.ru/artist/${artist.id}/tracks`,
          }));
      }
      return [];
    });
  }

  /**
   * Get random playlist, pick from USER_ID playlists
   * @return {PlaylistDto} playlists
   */
  async getRandomPlaylist(): Promise<PlaylistDto> {
    return firstValueFrom(
      this.httpService.get(`/users/${USER_ID_FOR_RANDOM}/playlists/list`, {}),
    ).then(({ data }) => {
      const filteredPlaylists = data.result.filter(
        (p) => p.revision && p.trackCount > 40,
      );
      const playlist =
        filteredPlaylists[randomIntByMax(filteredPlaylists.length)];
      return {
        id: playlist?.playlistUuid,
        name: playlist?.title,
        cover: 'https://' + playlist?.cover?.uri?.replace('%%', '200x200'),
        type: Type.playlist,
        url: `https://music.yandex.ru/users/${playlist?.owner?.login}/playlists/${playlist.kind}`,
      };
    });
  }

  /**
   * Get playlist by id
   * @param {string} id - playlist id
   * @param type
   * @return {PlaylistDto} playlist
   */
  async getPlaylistById(
    id: string,
    type: Type = Type.playlist,
  ): Promise<PlaylistDto> {
    return firstValueFrom(
      this.httpService.get(
        `/${type}${type === Type.playlist ? '' : 's'}/${id}`,
        {
          params: {
            fields: 'tracks',
          },
        },
      ),
    ).then(async ({ data }) => {
      if (type === Type.playlist) {
        return {
          id: data.result?.playlistUuid,
          name: data.result?.title,
          cover: 'https://' + data.result?.cover?.uri?.replace('%%', '200x200'),
          type,
          tracks: YandexService.parseTracks(
            data.result?.tracks.map(({ track }) => track),
          ),
          url: `https://music.yandex.ru/users/${data.result.owner.login}/playlists/${data?.result?.kind}`,
        };
      }
      if (type === Type.artist) {
        return {
          id: data.result.artist.id,
          name: data.result.artist.name,
          cover:
            'https://' +
            data.result.artist?.cover?.uri?.replace('%%', '200x200'),
          type,
          tracks: await this.getTracksByArtist(id),
          url: `https://music.yandex.ru/artist/${data.result.artist.id}/tracks`,
        };
      }
      return {
        id: undefined,
        name: undefined,
        cover: undefined,
        type,
        tracks: [],
        url: '',
      };
    });
  }

  /**
   * Get track by id
   * @param {string} id
   * @return {TrackDto} track
   */
  async getTrackById(id: string): Promise<TrackDto> {
    return firstValueFrom(this.httpService.get(`/tracks/${id}`)).then(
      async ({ data }) => {
        return {
          id: data.result?.[0]?.id,
          name: data.result?.[0]?.title,
          album: data.result?.[0]?.albums.map(({ title }) => title).join(', '),
          artist: data.result?.[0]?.artists.map(({ name }) => name).join(', '),
          mp3: await this.getMp3ByTrackId(id),
        };
      },
    );
  }

  async getMp3ByTrackId(id: string): Promise<any> {
    return firstValueFrom(
      this.httpService.get(`/tracks/${id}/download-info`),
    ).then(({ data }) =>
      firstValueFrom(
        this.httpService.get(data.result?.[0].downloadInfoUrl, {
          responseType: 'text',
        }),
      ).then(async (data) => {
        const { host, s, ts, path } = (
          await parseStringPromise(data.data, {
            trim: true,
            explicitArray: false,
          })
        )?.['download-info'];
        return `https://${host}/get-mp3/${s}/${ts}${path}`;
      }),
    );
  }

  /**
   * Get tracks by artist id
   * @param id
   */
  private async getTracksByArtist(id: string): Promise<TrackDto[]> {
    return firstValueFrom(
      this.httpService.get(`/artists/${id}/tracks`, {
        params: {
          page: 0,
          page_size: 1000,
        },
      }),
    ).then(({ data }) => YandexService.parseTracks(data.result.tracks));
  }

  private static parseTracks(tracks?: any[]): TrackDto[] {
    return (
      tracks?.map(({ albums, artists, id, title }) => ({
        id,
        name: title,
        artist: artists?.map(({ name }) => name).join(', '),
        album: albums?.map(({ title }) => title).join(', '),
      })) ?? []
    );
  }
}
