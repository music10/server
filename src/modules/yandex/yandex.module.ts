import { Module } from '@nestjs/common';
import { YandexHttpModule } from './yandexHttpModule';
import { YandexService } from './yandex.service';

@Module({
  imports: [YandexHttpModule],
  providers: [YandexService],
  exports: [YandexService],
})
export class YandexModule {}
