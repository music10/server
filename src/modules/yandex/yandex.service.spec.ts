import { Test, TestingModule } from '@nestjs/testing';
import { YandexService } from './yandex.service';
import { YandexHttpModule } from './yandexHttpModule';

describe('YandexApiService', () => {
  let service: YandexService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [YandexHttpModule],
      providers: [YandexService],
    }).compile();

    service = module.get<YandexService>(YandexService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should get playlists', async () => {
    expect(await service.getCherryPickPlaylists()).toMatchSnapshot();
  });

  it('should get featured playlists', async () => {
    expect((await service.getFeaturedPlaylists()).length).toBeGreaterThan(1);
  });

  it('should search playlists', async () => {
    expect(await service.searchPlaylists('русский')).toEqual(
      expect.arrayContaining([
        {
          id: expect.any(String),
          name: expect.any(String),
          cover: expect.any(String),
        },
      ]),
    );
  });

  it('should get tracks by artist', async () => {
    expect(await service.getTracksByArtist(158454)).toEqual(
      expect.arrayContaining([
        {
          album: expect.any(String),
          artist: expect.any(String),
          id: expect.any(String),
          mp3: expect.any(String),
          name: expect.any(String),
        },
      ]),
    );
  });

  it('should search artist', async () => {
    expect(await service.searchArtists('гуф')).toEqual(
      expect.arrayContaining([
        {
          id: expect.any(Number),
          name: expect.any(String),
          cover: expect.any(String),
        },
      ]),
    );
  });

  it('should get playlist by id', async () => {
    const result = await service.getPlaylistById(
      '901a037b-bfad-e30f-6325-6ebb49f412f3',
    );
    expect(result).toStrictEqual({
      cover:
        'https://avatars.yandex.net/get-music-user-playlist/70586/103372440.2102.42432/200x200?1558344142432',
      id: '901a037b-bfad-e30f-6325-6ebb49f412f3',
      name: 'Русский хип-хоп 2000-х',
    });
  });

  it('should get tracks by playlist id', async () => {
    const result = await service.getTracksByPlaylistId(
      '901a037b-bfad-e30f-6325-6ebb49f412f3',
    );
    expect(result.length).toBeGreaterThan(10);
    expect(result[0]).toStrictEqual({
      album: 'Громче воды, выше травы',
      artist: 'Каста',
      id: '597043',
      mp3: '',
      name: 'На порядок выше',
    });
  });

  it('should get track by id', async () => {
    const result = await service.getTrackById('597043');
    expect(result).toStrictEqual({
      artist: 'Каста',
      mp3: '',
      id: '597043',
      name: 'На порядок выше',
      album: 'Громче воды, выше травы',
    });
  });
});
