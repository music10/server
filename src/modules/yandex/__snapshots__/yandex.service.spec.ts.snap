// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YandexApiService should get cherry pick playlists 1`] = `
Array [
  Object {
    "cover": "https://musiq.dergunov.net/images/world.svg",
    "id": "7d5ba0bc-729e-1dd8-2df7-5da309b17c16",
    "name": "Мировые поп-хиты",
    "type": "playlist",
  },
  Object {
    "cover": "https://musiq.dergunov.net/images/russia.svg",
    "id": "e3e7e09a-f0d3-056a-8108-31f0a27368e8",
    "name": "Хиты русской поп-музыки",
    "type": "playlist",
  },
  Object {
    "cover": "https://musiq.dergunov.net/images/new.svg",
    "id": "2111e2b6-587d-a600-2fea-54df7c314477",
    "name": "Новинки месяца",
    "type": "playlist",
  },
  Object {
    "cover": "https://musiq.dergunov.net/images/tiktok.svg",
    "id": "c8c03392-31ee-730c-a8a6-c24a5a30e911",
    "name": "Топ тик-тока",
    "type": "playlist",
  },
  Object {
    "cover": "https://musiq.dergunov.net/images/hiphop.svg",
    "id": "479652a7-a86f-6770-3cbd-e1e0ac9600f1",
    "name": "Русский хип-хоп",
    "type": "playlist",
  },
  Object {
    "cover": "https://musiq.dergunov.net/images/rock.svg",
    "id": "b0a55e85-65c6-ce15-1478-6c693d6c525c",
    "name": "Вечный рок",
    "type": "playlist",
  },
]
`;
